### ISSUE TITLE
Cannot debug scan-build failure, scanbuild log is deleted

### ISSUE THREAD

**Issue by mcatanzaro** [mesonbuild/meson#8707]
**Describe the bug**

Epiphany's CI has started failing due to this strange scan-build issue:

```
$ flatpak build flatpak_app ninja -C _build scan-build
ninja: Entering directory `_build'
[0/1] /usr/bin/meson --internal scanbuild /home/mcatanzar...ize=4 -Dwarning_level=2 -Dwerror=true -Dwrap_mode=default
scan-build: Using '/usr/bin/clang-10' for static analysis
WARNING: Recommend using either -Dbuildtype or -Doptimization + -Ddebug. Using both is redundant since they override each other. See: https://mesonbuild.com/Builtin-options.html#build-type-options
The Meson build system
Version: 0.57.1
Source dir: /home/mcatanzaro/Projects/epiphany
Build dir: /home/mcatanzaro/Projects/epiphany/_build/meson-private/tmp8phqp91m
Build type: native build
Project name: epiphany
Project version: 40.0

../meson.build:1:0: ERROR: Compiler /usr/bin/../libexec/ccc-analyzer can not compile programs.

A full log can be found at /home/mcatanzaro/Projects/epiphany/_build/meson-private/tmp8phqp91m/meson-logs/meson-log.txt
scan-build: Removing directory '/tmp/scan-build-2021-04-27-150438-6-1' because it contains no reports.
scan-build: No bugs found.
FAILED: meson-scan-build 
/usr/bin/meson --internal scanbuild /home/mcatanzaro/Projects/epiphany /home/mcatanzaro/Projects/epiphany/_build /usr/bin/meson -Dauto_features=auto -Db_asneeded=true -Db_colorout=always -Db_coverage=false -Db_lto=false -Db_lto_threads=0 -Db_lundef=true -Db_ndebug=false -Db_pch=true -Db_pgo=off -Db_pie=false -Db_sanitize=none -Db_staticpic=true -Dbackend=ninja -Dbackend_max_links=0 -Dbindir=bin '-Dbuild.c_args=[]' '-Dbuild.c_link_args=['"'"'-l/app/lib'"'"']' -Dbuild.c_std=gnu11 '-Dbuild.cmake_prefix_path=[]' '-Dbuild.pkg_config_path=['"'"'/app/lib/pkgconfig'"'"', '"'"'/app/share/pkgconfig'"'"', '"'"'/usr/lib/pkgconfig'"'"', '"'"'/usr/share/pkgconfig'"'"']' -Dbuildtype=debug '-Dc_args=[]' '-Dc_link_args=['"'"'-l/app/lib'"'"']' -Dc_std=gnu11 '-Dcmake_prefix_path=[]' -Ddatadir=share -Ddebug=true -Ddefault_library=shared -Ddeveloper_mode=false -Derrorlogs=true '-Dforce_fallback_for=[]' -Dgsb_api_key= -Dincludedir=include -Dinfodir=share/info -Dinstall_umask=0o22 -Dlayout=mirror -Dlibdir=lib -Dlibexecdir=libexec -Dlibportal=enabled -Dlocaledir=share/locale -Dlocalstatedir=var -Dmandir=share/man -Dnetwork_tests=disabled -Doptimization=0 '-Dpkg_config_path=['"'"'/app/lib/pkgconfig'"'"', '"'"'/app/share/pkgconfig'"'"', '"'"'/usr/lib/pkgconfig'"'"', '"'"'/usr/share/pkgconfig'"'"']' -Dprefix=/app -Dprofile=Devel -Dsbindir=sbin -Dsharedstatedir=com -Dstdsplit=true -Dstrip=false -Dsysconfdir=etc -Dtech_preview=false -Dunit_tests=enabled -Dunity=off -Dunity_size=4 -Dwarning_level=2 -Dwerror=true -Dwrap_mode=default
ninja: build stopped: subcommand failed.
```

Note: `../meson.build:1:0: ERROR: Compiler /usr/bin/../libexec/ccc-analyzer can not compile programs.` ~~I don't have any reason to believe the failure is itself a meson bug, but I'm having trouble debugging it because meson is eating the debug log.~~ /home/mcatanzaro/Projects/epiphany/_build/meson-private/tmp8phqp91m does not exist: it's created and then immediately deleted [here in scanbuild.py](https://github.com/mesonbuild/meson/blob/a493c950a73c5924e1e2f9454a5c13d0c7bc93bc/mesonbuild/scripts/scanbuild.py#L26) by python's tempfile.TemporaryDirectory. It's deleted automatically at the end of the `with` clause.

Honestly, looking at the code in scanbuild.py, it looks like the log file is probably *intended* to go directly under the toplevel meson-logs, rather than under meson-private/tmpdir/meson-logs, but that is not what's actually happening.

**To Reproduce**
[Instructions to reproduce here.](https://gitlab.gnome.org/GNOME/epiphany/-/merge_requests/953#note_1090064) It's a little complicated reproducer since it involves `flatpak build`, but that allows reproducing with a consistent toolchain without any host dependencies.

**Expected behavior**
Meson says "A full log can be found at /home/mcatanzaro/Projects/epiphany/_build/meson-private/tmp8phqp91m/meson-logs/meson-log.txt" so I would expect that directory and log to exist.

**system parameters**
* native build
* freedesktop-sdk 20.08.8
* python 3.8.8
* meson 0.57.1
* ninja 1.10.2

**Comment by alatiera** [comment]
71db6b0 is the last working commit. I've bisected this to e81acbd6069e8c1ae8e3be7bb83ddc239009d42d

It was weird cause this commit caused at least two issues and the bisect ended up being painful. First issue has been fixed by #8282. 

```
../meson.build:1:0: ERROR: Invalid mode: invalid literal for int() with base 8: '18'
```


After cherry-picking f31ffaaf1754e4578127049844c14eba6bdda477 we get back to the `ccc-analyzer` issue.

```
ninja: Entering directory `build'
[0/1] /usr/bin/python3 /home/alatiera/Projects/meson/meson.py --i...unity_size=4 -Dwarning_level=2 -Dwerror=false -Dwrap_mode=default
scan-build: Using '/usr/bin/clang-10' for static analysis
WARNING: Recommend using either -Dbuildtype or -Doptimization + -Ddebug. Using both is redundant since they override each other. See: https://mesonbuild.com/Builtin-options.html#build-type-options
The Meson build system
Version: 0.56.99
Source dir: /home/alatiera/Projects/epiphany
Build dir: /home/alatiera/Projects/epiphany/build/meson-private/tmph6kzjijz
Build type: native build
Project name: epiphany
Project version: 40.0

../meson.build:1:0: ERROR: Compiler /usr/bin/../libexec/ccc-analyzer can not compile programs.

A full log can be found at /home/alatiera/Projects/epiphany/build/meson-private/tmph6kzjijz/meson-logs/meson-log.txt
scan-build: Removing directory '/tmp/scan-build-2021-04-28-032145-115292-1' because it contains no reports.
scan-build: No bugs found.
FAILED: meson-scan-build 
/usr/bin/python3 /home/alatiera/Projects/meson/meson.py --internal scanbuild /home/alatiera/Projects/epiphany /home/alatiera/Projects/epiphany/build /usr/bin/python3 /home/alatiera/Projects/meson/meson.py -Dauto_features=auto -Db_asneeded=true -Db_colorout=always -Db_coverage=false -Db_lto=false -Db_lundef=true -Db_ndebug=false -Db_pch=true -Db_pgo=off -Db_pie=false -Db_sanitize=none -Db_staticpic=true -Dbackend=ninja -Dbackend_max_links=0 -Dbindir=bin '-Dbuild.c_args=[]' '-Dbuild.c_link_args=['"'"'-l/app/lib'"'"']' -Dbuild.c_std=gnu11 '-Dbuild.cmake_prefix_path=[]' '-Dbuild.pkg_config_path=['"'"'/app/lib/pkgconfig'"'"', '"'"'/app/share/pkgconfig'"'"', '"'"'/usr/lib/pkgconfig'"'"', '"'"'/usr/share/pkgconfig'"'"']' -Dbuildtype=debug '-Dc_args=[]' '-Dc_link_args=['"'"'-l/app/lib'"'"']' -Dc_std=gnu11 '-Dcmake_prefix_path=[]' -Ddatadir=share -Ddebug=true -Ddefault_library=shared -Ddeveloper_mode=false -Derrorlogs=true '-Dforce_fallback_for=[]' -Dgsb_api_key= -Dincludedir=include -Dinfodir=share/info -Dinstall_umask=0o22 -Dlayout=mirror -Dlibdir=lib64 -Dlibexecdir=libexec -Dlibportal=enabled -Dlocaledir=share/locale -Dlocalstatedir=/var/local -Dmandir=share/man -Dnetwork_tests=disabled -Doptimization=0 '-Dpkg_config_path=['"'"'/app/lib/pkgconfig'"'"', '"'"'/app/share/pkgconfig'"'"', '"'"'/usr/lib/pkgconfig'"'"', '"'"'/usr/share/pkgconfig'"'"']' -Dprefix=/usr/local -Dprofile= -Dsbindir=sbin -Dsharedstatedir=/var/local/lib -Dstdsplit=true -Dstrip=false -Dsysconfdir=etc -Dtech_preview=false -Dunit_tests=enabled -Dunity=off -Dunity_size=4 -Dwarning_level=2 -Dwerror=false -Dwrap_mode=default
ninja: build stopped: subcommand failed.
bash-5.1$ cat /home/alatiera/Projects/epiphany/build/meson-private/tmph6kzjijz/meson-logs/meson-log.txt
cat: /home/alatiera/Projects/epiphany/build/meson-private/tmph6kzjijz/meson-logs/meson-log.txt: No such file or directory
```

Any ideas where we can get any more logs about it? Anything else I can provide?

**Comment by alatiera** [comment]
Oh whoops, I just noticed this issues is about the logs being deleted not the `ccc-analyzer` bug, guess we should move that over to a new one.

**Comment by mcatanzaro** [comment]
> [71db6b0](https://github.com/mesonbuild/meson/commit/71db6b04a31707674ad776be1cf22f667056d56b) is the last working commit. I've bisected this to [e81acbd](https://github.com/mesonbuild/meson/commit/e81acbd6069e8c1ae8e3be7bb83ddc239009d42d)

Wow, thanks!

Oddly, this bug does NOT occur if I install the exact same meson version, meson 0.57.1, on my host system....

**Comment by mcatanzaro** [comment]
I've split the actual build failure out into #8818, so we can use this bug just to track the logging problem.

Note we've also downgraded meson in freedesktop-sdk in order to unbreak `meson scan-build` there.

**Comment by mcatanzaro** [comment]
Fixed by #8823.

(#8818 remains open.)
